### PR TITLE
feat: Implement custom bootstrapping on startup

### PR DESCRIPTION
**Pull request by sighphyre**
Adds the ability to bootstrap Toggle states from a custom source on load of the client. Currently, there's direct support for loading from a URL and/or loading from a file

### CODE REVIEW

**Comment by sighphyre** [comment]
This is related to https://github.com/Unleash/unleash-client-ruby/issues/64

This adds support for loading from a file,URL or custom implementation

File: lib/unleash/bootstrap.rb [hunk 1]
```diff
@@ -0,0 +1,55 @@
+require 'unleash/configuration'
+require 'unleash/feature_toggle'
+require 'logger'
+require 'time'
+require 'net/http'
+require 'uri'
+
+module Unleash
```
Comment (ivarconr) [inline lib/unleash/bootstrap.rb:8]: It would be nice to also see a `data` bootstrapper out of the box.

File: lib/unleash/toggle_fetcher.rb [hunk 2]
```diff
@@ -4,20 +4,26 @@
...
       self.toggle_cache = nil
       self.toggle_lock = Mutex.new
       self.toggle_resource = ConditionVariable.new
       self.retry_count = 0
+      self.bootstrapper = Unleash.configuration.bootstrapper
 
       # start by fetching synchronously, and failing back to reading the backup file.
       begin
-        fetch
+        if !self.bootstrapper.nil? # if the consumer provides a bootstrapper, we're going to assume they want to use it
```
Comment (ivarconr) [inline lib/unleash/toggle_fetcher.rb:19]: I am unsure if I understand this code block. We want to:
1. Load configuration from the boostrapper 
2. Load configuration from the Unleash API (fetch) as soon as possible.

File: lib/unleash/bootstrap.rb [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    def build_headers(headers = nil)
+      headers = (headers || {}).dup
+      headers['Content-Type'] = 'application/json'
+
+      headers
+    end
+  end
+
+  def self.extract_bootstrap(bootstrap_hash)
+    raise NotImplemented, "The provided bootstrap doesn't seem to be a valid set of toggles" if bootstrap_hash['version'] < 1
```
Comment (ivarconr) [inline lib/unleash/bootstrap.rb:51]: I am unsure if we want to require the version field here.

**Review by rarruda — COMMENTED** [review]
I would like to see Bootstrap as a separate Module, with different classes in it, each providing an implementation of an abstract class.

Note that ideally each class should be in a separate file.

For example please look at https://github.com/Unleash/unleash-client-ruby/blob/main/lib/unleash/strategy/base.rb and https://github.com/Unleash/unleash-client-ruby/blob/main/lib/unleash/strategy/default.rb

If you want/have the time/patience, I can try creating a branch on top of yours to show what I mean, but It could take me a couple of days until i can get around to it.

File: lib/unleash/bootstrap.rb [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+  class UrlBootStrapper
+    attr_accessor :uri, :headers
+
+    def initialize(uri, headers)
+      self.uri = URI(uri)
+      self.headers = headers
+    end
+
+    def read
+      request = Net::HTTP::Get.new(self.uri, self.build_headers(self.headers))
```
Comment (rarruda) [inline lib/unleash/bootstrap.rb:32]: Note that as a "Built-in" class, then it should use the functionalities from Unleash::Util::Http

File: lib/unleash/bootstrap.rb (hunk 1 above)
Comment (rarruda) [reply lib/unleash/bootstrap.rb]: So, this is now possible after #86 got merged.

The main concern is that the data would need to be serialized in JSON first (to only then be de-serialized).

Having an extra `Unleash::Bootstrap::FromData` class doesn't seem like a bad idea though.

File: lib/unleash/bootstrap.rb (hunk 4 above)
Comment (rarruda) [reply lib/unleash/bootstrap.rb]: Fixed after the merge of #86

File: README.md [hunk 5]
```diff
@@ -267,6 +268,27 @@ variant = UNLEASH.get_variant "ColorVariants", @unleash_context, fallback_varian
...
+
+`bootstrap_data` configuration allows the client to be initialized with a predefined set of toggle states.
+The content of the parameter is a JSON string containing the response body from the unleash server.
+
+We provide two classes to help fetch the bootstrap files:
+* `Unleash::Bootstrap::FromFile`
+* `Unleash::Bootstrap::FromUri`
+
+Example usage:
+```ruby
```
Comment (rarruda) [inline README.md:281]:
```suggestion

First saving the toggles locally:
```shell
curl -H 'Authorization: abc' -XGET 'http://unleash.herokuapp.com/api' > ./default-toggles.json
```

Now using them on start up:
```ruby
```

File: README.md [hunk 6]
```diff
@@ -305,6 +327,12 @@ This client comes with the all the required strategies out of the box:
  * UnknownStrategy
  * UserWithIdStrategy
 
+## Available Bootstrap Classes
+
+This client comes with these classes to load unleash features on startup, before making a request to the Unleash API:
+
+ * Unleash::Bootstrap::FromFile
+ * Unleash::Bootstrap::FromUri
```
Comment (rarruda) [inline README.md:335]: Maybe add examples, including how to set HTTP headers.

File: examples/default-toggles.json [hunk 7]
```diff
@@ -0,0 +1,43 @@
+
```
Comment (rarruda) [inline examples/default-toggles.json:1]:
```suggestion
```

File: examples/default-toggles.json [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+            {
+                "name": "hola",
+                "parameters": {
+                    "name": "val"
+                }
+            }
+        ]
+
+    }
+]}
```
Comment (rarruda) [inline examples/default-toggles.json:43]:
```suggestion
]
}
```

File: spec/unleash/bootstrap/from_uri_spec.rb [hunk 9]
```diff
@@ -0,0 +1,27 @@
+require 'unleash/bootstrap/from_uri'
+require 'json'
+
+RSpec.describe Unleash::Bootstrap::FromUri do
+  it 'loads bootstrap toggle correctly from URL' do
+    bootstrap_file = './spec/unleash/bootstrap-resources/features-v1.json'
```
Comment (rarruda) [inline spec/unleash/bootstrap/from_uri_spec.rb:6]: maybe consider in-lining the JSON here. The data is not really re-used in different contexts.

File: lib/unleash/toggle_fetcher.rb (hunk 2 above)
Comment (rarruda) [reply lib/unleash/toggle_fetcher.rb]: Is it ok to wait for the next update cycle? Or should that happen immediately after loading via the bootstraping mechanism?

File: lib/unleash/bootstrap.rb (hunk 3 above)
Comment (rarruda) [reply lib/unleash/bootstrap.rb]: In the merged (from #86) we use the same test for the fetched toggles as we do for the bootstrapped ones. Which seems pretty reasonable imho.

File: README.md [hunk 10]
```diff
@@ -267,6 +268,27 @@ variant = UNLEASH.get_variant "ColorVariants", @unleash_context, fallback_varian
...
+#### Bootstrapping
+
+`bootstrap_data` configuration allows the client to be initialized with a predefined set of toggle states.
+The content of the parameter is a JSON string containing the response body from the unleash server.
+
+We provide two classes to help fetch the bootstrap files:
+* `Unleash::Bootstrap::FromFile`
+* `Unleash::Bootstrap::FromUri`
+
+Example usage:
```
Comment (rarruda) [inline README.md:280]:
```suggestion
Example usage:

First saving the toggles locally:
```shell
curl -H 'Authorization: abc' -XGET 'http://unleash.herokuapp.com/api' > ./default-toggles.json
```.

Now using them on start up:
```

**Pull request closed without merging**